components/service/glean/src/main/java/mozilla/components/service/glean/histogram/PrecomputedHistogram.kt

```diff
@@ -0,0 +1,276 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.histogram
+
+import mozilla.components.service.glean.private.HistogramType
+import mozilla.components.support.ktx.android.org.json.tryGetInt
+import mozilla.components.support.ktx.android.org.json.tryGetLong
+import mozilla.components.support.ktx.android.org.json.tryGetString
+import org.json.JSONArray
+import org.json.JSONObject
+
+/**
+ * This class represents the structure of a custom distribution. It is meant
+ * to help serialize and deserialize data to the correct format for transport and
+ * storage, as well as including helper functions to calculate the bucket sizes.
+ *
+ * @param rangeMin the minimum value that can be represented
+ * @param rangeMax the maximum value that can be represented
+ * @param bucketCount total number of buckets
+ * @param histogramType the [HistogramType] representing the bucket layout
+ * @param values a map containing the bucket index mapped to the accumulated count
+ * @param sum the accumulated sum of all the samples in the custom distribution
+ */
+data class PrecomputedHistogram(
```

Comment:
Bikeshedding: what if we'll need to precompute `bucketIndexToBucketMinimum` for the `FunctionalHistogram` due to performance reasons? Would it make the latter pre-computed too?

Reply:
In that case, I would use the `PrecomputedHistogram` implementation and add a new `histogram_type` enum to handle the different bucketing (probably with some feature so we only have a single static buckets list, since it will be much larger than the other ones).

I'm skeptical that the binary search will beat out the log calculation, though, since the former is memory bound...  But all that is conjecture without benchmarking, of course.